File: persistence/src/main/java/org/panteleyev/money/persistence/MySQLBuilder.java
package org.panteleyev.money.persistence;

/*
 Copyright (c) Petr Panteleyev. All rights reserved.
 Licensed under the BSD license. See LICENSE file in the project root for full license information.
 */

import com.mysql.cj.jdbc.MysqlDataSource;
import javax.sql.DataSource;
import java.sql.SQLException;
import java.util.TimeZone;

public class MySQLBuilder {
    private int port = 3306;
    private String host = "localhost";
    private String dbName = "";
    private String user = "";
    private String password = "";

    public DataSource build() {
        try {
            var ds = new MysqlDataSource();
            ds.setCharacterEncoding("utf8");
            ds.setUseSSL(false);
            ds.setServerTimezone(TimeZone.getDefault().getID());
            ds.setPort(port);
            ds.setServerName(host);
            ds.setUser(user);
            ds.setPassword(password);
            ds.setDatabaseName(dbName);
            return ds;
        } catch (SQLException ex) {
            throw new RuntimeException(ex);
        }
    }

    public MySQLBuilder host(String host) {
        this.host = host;
        return this;
    }

    public MySQLBuilder port(int port) {
        this.port = port;
        return this;
    }

    public MySQLBuilder name(String name) {
        this.dbName = name;
        return this;
    }

    public MySQLBuilder user(String user) {
        this.user = user;
        return this;
    }

    public MySQLBuilder password(String password) {
        this.password = password;
        return this;
    }

    public String getConnectionString() {
        return "mysql://" + host + ":" + port + "/" + dbName;
    }
}

File: src/main/kotlin/org/panteleyev/money/persistence/MySQLBuilder.kt
/*
 * Copyright (c) 2017, Petr Panteleyev <petr@panteleyev.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package org.panteleyev.money.persistence

import com.mysql.jdbc.jdbc2.optional.MysqlDataSource
import javax.sql.DataSource

class MySQLBuilder {
    var port = 3306
    var host = "localhost"
    var dbName = ""
    var user = ""
    var password = ""

    fun build() : DataSource {
        val ds : MysqlDataSource = MysqlDataSource()
        ds.encoding = "utf8"
        ds.port = port
        ds.serverName = host
        ds.user = user
        ds.setPassword(password)
        ds.databaseName = dbName
        return ds
    }

    fun host(host : String) = apply { this.host = host }
    fun port(port : Int) = apply { this.port = port }
    fun name(name : String) = apply { this.dbName = name }
    fun user(user : String) = apply { this.user = user }
    fun password(password : String) = apply { this.password = password }

    fun connectionString() : String = "mysql://$host:$port/$dbName"
}